### DISCUSSION TITLE
Numpy array to Vectorarray using from_numpy()

### DISCUSSION THREAD

**Discussion by shashimallapur** [pymor/pymor#1629] [Q&A] [answered]
Hi All,

I am having problems trying to convert a numpy array (which contains the sample data) to Vectorarray datatype. Below you can find the code snippet and I have  attached the error as well.

I intend to use this Vectorarray for the dmd algorithm. Thank you all :)

![Error_from_numpy()](https://user-images.githubusercontent.com/105507935/168320890-b239f649-a89f-4e31-9ed3-1c8f0c04fbe6.PNG)

```python
import pandas as pd
import numpy as np
from scipy.linalg import svd
from pymor.basic import*

from pymor.vectorarrays.interface import VectorSpace

X_df =pd.read_excel('data_vel_X.xlsx') # importing the sample data as data frame
X = X_df.to_numpy() # converting data to numpy array

X_obj=VectorSpace # initiating object of class Vectorspace
X_obj.from_numpy(X,ensure_copy=False) # converting data as numpy array to Vectorarray
```

Edit (@renefritze): Made the snippet a code block

**Comment by pmli** [comment, marked as answer]
Hi @shashimallapur,

thank you for the question.

The idea in pyMOR is that `VectorSpace` defines a generic interface, while its subclasses give concrete implementations. For NumPy, this would be the `NumpyVectorSpace` (some more can be found in `pymor.bindings`). Assuming that the columns of `X` are the snapshots at particular times, the following should work

```python
from pymor.vectorarrays.numpy import NumpyVectorSpace

dim = X.shape[0]
X_va = NumpyVectorSpace(dim).from_numpy(X.T)
```

**Reply by shashimallapur** [reply]
Hi @pmli,

thank you very much for the prompt reply and the thorough explanantion.
Your code snippet works perfectly!

Thanks a lot :)

**Reply by shashimallapur** [reply]
Hi @pmli,

I have another question regarding Vectorarrays. I tried to convert the Vectorarray back to numpy array using to_numpy function, in a similar way as you showed me earlier ;) Unfortunately I am getting the following error. Would be grateful if you can help me with this one as well :)

Code snippet:

from pymor.vectorarrays.numpy import NumpyVectorSpace
from pymor.vectorarrays.numpy import NumpyVectorArray
![Error_from_numpy()](https://user-images.githubusercontent.com/105507935/169370146-c0e54207-a2a4-4861-b018-d028069a44b1.PNG)


dim = X.shape[0]
X_va = NumpyVectorSpace(dim).from_numpy(X.T)

A_tilde_X=dmd(X_va)

A_tilde_X_num=NumpyVectorArray(dim).to_numpy(A_tilde_X)

**Reply by pmli** [reply]
Hi, there are two things here:

1. [`dmd`](https://docs.pymor.org/2021-2-0/autoapi/pymor/algorithms/dmd/index.html#pymor.algorithms.dmd.dmd) returns a tuple (not just the `VectorArray`),
2. `to_numpy` should be used as an instance method.

The following should work

```python
W_va, evals = dmd(X_va)
W = W_va.to_numpy().T
```

**Reply by shashimallapur** [reply]
Thanks a lot for the solution @pmli 
It works perfect :)
